db/dcrpg/queries.go

```diff
@@ -3619,6 +3620,182 @@ func appendBlockFees(charts *cache.ChartData, rows *sql.Rows) error {
 	return rows.Err()
 }
 
+// retrievePrivacyParticipation retrieves the sum of all mixed vouts that is newer than the data
+// in the provided ChartData. This data is used to plot fees on the /charts page.
+// This is the Fetcher half of a pair that make up a cache.ChartUpdater.
+func retrievePrivacyParticipation(ctx context.Context, db *sql.DB, charts *cache.ChartData) (*sql.Rows, error) {
+	rows, err := db.QueryContext(ctx, internal.SelectMixedTotalPerBlock, charts.TotalMixedTip())
+	if err != nil {
+		return nil, err
+	}
+	return rows, nil
+}
+
+// Append the result from retrievePrivacyParticipation to the provided ChartData. This
+// is the Appender half of a pair that make up a cache.ChartUpdater.
+func appendPrivacyParticipation(charts *cache.ChartData, rows *sql.Rows) error {
+	defer rows.Close()
+	blocks := charts.Blocks
+	for rows.Next() {
+		var blockHeight uint64
+		var totalMixed int64
+		if err := rows.Scan(&blockHeight, &totalMixed); err != nil {
+			log.Errorf("Unable to scan for MixedVoutsPerBlock fields: %v", err)
+			return err
+		}
+		if totalMixed < 0 {
+			totalMixed *= -1
+		}
+
+		// Converting to atoms.
+		blocks.TotalMixed = append(blocks.TotalMixed, uint64(totalMixed))
+	}
+	return rows.Err()
+}
+
+// retrieveAnonymitySet retrieves any block total mixed data that is newer than the data
+// in the provided ChartData. This data is used to plot anonymity-set on the /charts page.
+// This is the Fetcher half of a pair that make up a cache.ChartUpdater.
+func retrieveAnonymitySet(ctx context.Context, db *sql.DB, _ *cache.ChartData) (*sql.Rows, error) {
+	rows, err := db.QueryContext(ctx, internal.SelectMixedVouts)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+// Append the result from retrieveAnonymitySet to the provided ChartData. This
+// is the Appender half of a pair that make up a cache.ChartUpdater.
+func appendAnonymitySet(charts *cache.ChartData, rows *sql.Rows) (err error) {
+	var vals, fundHeights, spendHeights []int64
+
+	var maxHeight int64
+	minHeight := int64(math.MaxInt64)
+
+
+	for rows.Next() {
+		var value, fundHeight, spendHeight int64
+		var spendHeightNull sql.NullInt64
+		var tree uint8
+		err = rows.Scan(&value, &fundHeight, &spendHeightNull, &tree)
+		if err != nil {
+			return err
+		}
+
+		if spendHeightNull.Valid {
+			spendHeight = spendHeightNull.Int64
+		} else {
+			spendHeight = -1
+		}
+
+		if fundHeight < minHeight {
+			minHeight = fundHeight
+		}
+
+		if spendHeight > maxHeight {
+			maxHeight = spendHeight
+		}
+
+		vals = append(vals, value)
+		fundHeights = append(fundHeights, fundHeight)
+		spendHeights = append(spendHeights, spendHeight)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Errorf("Unable to scan for BlockCoinJoins fields: %v", err)
+	}
+
+	blocks := charts.Blocks
+
+	// block anonymity set
+	for _, h := range blocks.Height[len(blocks.AnonymitySet) :] {
```

Comment:
Is there a guarantee that `blocks.Height` will be at least this long?  This will inevitably panic and crash dcrdata when something goes wrong.

Why not infer the block height like the other appenders, and loop like:

```go
for h := len(blocks.AnonymitySet)+1; h <= len(blocks.Height); h++ {
	...
}
```

And/or first check the length of blocks.Height to catch malfunctions.